### PR TITLE
Shakemap functions

### PR DESCRIPTION
**Pull request by schmidni**
My results of trying to improve performance of the correlation calculations for shakemaps.

* changes in cross_corr_matrix are mostlya changing of the if/else orders
* spatial_covariance_array had a matrix multiplication using for loops, doing a proper matmul using np had a major improvement in performance
* cholesky function is the one that takes now the longest. I tried switching out all loops with matrix manipulations but could not get a noticeable improvement. Apart from the cholesky decomposition, the L * L.T obviously just takes a fixed amount of time.

### CODE REVIEW

**Comment by schmidni** [comment]
Also, there are two more "direct" ways to calculate the LLT matrix in the cholesky function, which however give no performance improvement:

using:
`    L = numpy.array([numpy.linalg.cholesky(spatial_cov[i])
                     for i in range(M)]).reshape(N*M, N)`

it is possible to get LLT 

`    LLT = (numpy.matmul(L, L.T).reshape(M, N, M, N)
           * cross_corr[:, None, :, None]).reshape(N*M, N*M)`
          
or

`    LLT = numpy.multiply(
        L@L.T, numpy.kron(cross_corr, numpy.ones((N, N))))`

Without any direct performance or readability improvement I didn't see a reason to include those changes. I can include them however if you want.

**Comment by micheles** [comment]
I like this. BTW, I am not the author of the code that you refactored ;-)

File: openquake/hazardlib/shakemap.py [hunk 1]
```diff
@@ -198,25 +197,24 @@ def cross_correlation_matrix(imts, corr='yes'):
...
-        T1 = im.period or 0.05
-
-        for j in range(M):
-            T2 = imts[j].period or 0.05
-            if i == j:
-                cross_matrix[i, j] = 1
-            else:
+    cross_matrix = numpy.eye(M)
+    if corr == 'full':
+        cross_matrix = numpy.full((M, M), 0.99999)
```
Comment (micheles) [inline openquake/hazardlib/shakemap.py:202]: We would need a comment here explaining the magic number  0.99999

Comment (schmidni) [reply openquake/hazardlib/shakemap.py:202]: Took the magic number from line 214 of the original code. 

I can only guess that there is some numerical reason?

Comment (micheles) [reply openquake/hazardlib/shakemap.py:202]: You are right, it was there from the beginning. Let's me accept the PR as it is, then.

**Review by micheles — APPROVED** [review]

**Pull request merged**